c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -15,25 +20,123 @@ mod dump;
+}
+
+/// Metadata describing the lifetime parameters and fields
+/// of a struct.
+#[derive(Clone, PartialEq, Eq)]
+pub struct AdtMetadata<'tcx> {
+    /// The lifetime parameters of a structure, including
+    /// hypothetical lifetimes derived from pointer fields.
+    pub lifetime_params: IndexSet<OriginParam>,
+    pub field_info: IndexMap<DefId, FieldMetadata<'tcx>>,
+}
+
+impl Default for AdtMetadata<'_> {
+    fn default() -> Self {
+        Self {
+            lifetime_params: IndexSet::new(),
+            field_info: IndexMap::new(),
+        }
+    }
+}
+
+/// An origin parameter of a type to resolve in a MIR body
+/// that will get mapped to a concrete Origin to
+/// provide to polonius.
+#[derive(Hash, PartialEq, Eq, Clone, Copy)]
+pub enum OriginParam {
+    /// An existing [`EarlyBoundRegion`], i.e. `'a` in `struct A<'a>`
+    Actual(EarlyBoundRegion),
+    /// A hypothesized region derived from a pointer type
+    /// e.g. `'h0` derived from the pointer in `*mut foo`
+    Hypothetical(i64),
+}
+
+/// An origin parameter of a field type resolve in a MIR body
+/// that will get mapped to a concrete Origin to
+/// provide to polonius.
+#[derive(Hash, PartialEq, Eq, Clone, Copy)]
+pub enum OriginArg<'tcx> {
+    /// An existing [`Region`], i.e. `'a` in `&'a foo`.
+    /// Can be [RegionKind::ReEarlyBound](`rustc_type_ir::RegionKind`) or [RegionKind::ReStatic](`rustc_type_ir::RegionKind`)
+    Actual(Region<'tcx>),
+    /// A hypothesized region derived from a pointer type
+    /// e.g. `'h0` derived from the pointer in `*mut foo`
+    Hypothetical(i64),
+}
+
+impl<'tcx> std::fmt::Debug for OriginArg<'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self {
+            OriginArg::Actual(r) => write!(f, "{:}", r),
+            OriginArg::Hypothetical(h) => write!(f, "'h{h:?}"),
+        }
+    }
+}
+
+impl TryFrom<&OriginArg<'_>> for OriginParam {
+    type Error = ();
+    fn try_from(value: &OriginArg<'_>) -> Result<Self, Self::Error> {
+        Ok(match value {
+            OriginArg::Hypothetical(h) => OriginParam::Hypothetical(*h),
+            OriginArg::Actual(r) => {
+                if let ReEarlyBound(eb) = r.kind() {
+                    OriginParam::Actual(eb)
+                } else {
+                    return Err(());
+                }
+            }
+        })
+    }
+}
+
+impl std::fmt::Debug for OriginParam {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self {
+            OriginParam::Actual(r) => write!(f, "{:}", r.name),
+            OriginParam::Hypothetical(h) => write!(f, "'h{h:?}"),
+        }
+    }
+}
 
 pub fn borrowck_mir<'tcx>(
     acx: &AnalysisCtxt<'_, 'tcx>,
     dataflow: &DataflowConstraints,
     hypothesis: &mut PointerTableMut<PermissionSet>,
     name: &str,
     mir: &Body<'tcx>,
+    adt_metadata: &AdtMetadataTable<'tcx>,
+    field_tys: HashMap<DefId, crate::LTy<'tcx>>,
```

Comment:
Is it useful enough to name both `LabeledTy<'tcx, PointerId>` and `LabeledTy<'tcx, Label<'tcx>>` as `LTy<'tcx>`?

Reply:
What do you mean by useful enough?

Reply:
Sorry, that was less clear than in my head.  I mean, naming them both `LTy` is pretty confusing, but if `LTy` is used so ubiquitously in its own module that the short name is quite useful, then it could be a better choice still.

Reply:
The `LabeledTy<Label>` type is used pervasively in the `borrowck` module, so it should have a short name, but I think it would be a good idea to rename it (in a separate PR) to something like `BTy` ("borrowck type") so it can't be confused with the `LTy = LabeledTy<PointerId>` type.